File: character-sheet-companion-server/src/listeners/actor/actorAttributesListener.ts
import { Server, Socket } from "socket.io";
import {
  getFoundrySocketFromRoom,
  getIosSocketFromRoom,
} from "../../utilities/SocketUtilities";
import { SOCKET_EVENTS } from "../../constants/events";
import { ActorAttributes } from "../../interfaces/actor/ActorAttributes";

export const actorAttributesListener = (io: Server, socket: Socket): void => {
  const requestActorAttributes = (actorId: string) => {
    const foundrySocket = getFoundrySocketFromRoom(io, socket);
    if (foundrySocket) {
      foundrySocket.emit(
        SOCKET_EVENTS.SERVER.ACTOR.REQUEST.REQUEST_ACTOR_ATTRIBUTES,
        actorId,
        socket.id
      );
    }
  };

  const receiveActorAttributes = (
    actorAttributes: ActorAttributes,
    iosSocketId: string
  ) => {
    const iosSocket = getIosSocketFromRoom(io, socket, iosSocketId);
    if (iosSocket) {
      iosSocket.emit(
        SOCKET_EVENTS.SERVER.ACTOR.SEND.SEND_ACTOR_ATTRIBUTES,
        JSON.stringify(actorAttributes)
      );
    }
  };

  const requestActorHealthChange = (actorId: string, healthChange: number) => {
    const foundrySocket = getFoundrySocketFromRoom(io, socket);
    if (foundrySocket) {
      foundrySocket.emit(
        SOCKET_EVENTS.SERVER.ACTOR.REQUEST.REQUEST_ACTOR_HEALTH_CHANGE,
        actorId,
        healthChange,
        socket.id
      );
    }
  };

  const requestShortRest = (actorId: string) => {
    const foundrySocket = getFoundrySocketFromRoom(io, socket);
    if (foundrySocket) {
      foundrySocket.emit(
        SOCKET_EVENTS.SERVER.ACTOR.REQUEST.REQUEST_ACTOR_SHORT_REST,
        actorId,
        socket.id
      );
    }
  };

  const requestLongRest = (actorId: string) => {
    const foundrySocket = getFoundrySocketFromRoom(io, socket);
    if (foundrySocket) {
      foundrySocket.emit(
        SOCKET_EVENTS.SERVER.ACTOR.REQUEST.REQUEST_ACTOR_LONG_REST,
        actorId,
        socket.id
      );
    }
  };

  socket.on(
    SOCKET_EVENTS.IOS.ACTOR.REQUEST_ACTOR_ATTRIBUTES,
    requestActorAttributes
  );
  socket.on(
    SOCKET_EVENTS.FOUNDRY.ACTOR.SEND_ACTOR_ATTRIBUTES,
    receiveActorAttributes
  );
  socket.on(
    SOCKET_EVENTS.IOS.ACTOR.REQUEST_ACTOR_HEALTH_CHANGE,
    requestActorHealthChange
  );
  socket.on(SOCKET_EVENTS.IOS.ACTOR.REQUEST_ACTOR_SHORT_REST, requestShortRest);
  socket.on(SOCKET_EVENTS.IOS.ACTOR.REQUEST_ACTOR_LONG_REST, requestLongRest);
};

File: CharacterSheetCompanion/CharacterSheetCompanion/Services/SocketListeners/ActorListeners/ActorAttributesListener.swift
//
//  ActorOverviewListener.swift
//  CharacterSheetCompanion
//
//  Created by Zachary Sanders on 8/11/21.
//

import Combine
import Foundation
import SocketIO

class ActorAttributesListener: SocketListener, ActorListener {
    let attributesPublisher: AnyPublisher<ActorAttributesModel?, Never>
    let socket: SocketIOClient

    private var receivedFirstMessage = false
    private let attributesSubject = CurrentValueSubject<ActorAttributesModel?, Never>(nil)

    init(socket: SocketIOClient) {
        attributesPublisher = attributesSubject.eraseToAnyPublisher()
        self.socket = socket
    }

    func addSocketHandlers() {
        socket.on(SocketEvents.SERVER.ACTOR.SEND.SEND_ACTOR_ATTRIBUTES) { data, _ in
            do {
                try self.attributesSubject.send(SocketListenerUtility.parseSocketEventData(data))
                self.receivedFirstMessage = true
            } catch let FoundryJSONError.errorMessage(errorMessage) {
                print(errorMessage)
            } catch {
                print(error)
            }
        }
    }

    func requestActorHealthChange(healthChange: Int) {
        if let actor = SocketManagerWrapper.sharedInstance.actor {
            socket.emit(SocketEvents.IOS.ACTOR.REQUEST_ACTOR_HEALTH_CHANGE, actor.id, healthChange)
        }
    }

    func requestActorShortRest() {
        if let actor = SocketManagerWrapper.sharedInstance.actor {
            socket.emit(SocketEvents.IOS.ACTOR.REQUEST_ACTOR_SHORT_REST, actor.id)
        }
    }

    func requestActorLongRest() {
        if let actor = SocketManagerWrapper.sharedInstance.actor {
            socket.emit(SocketEvents.IOS.ACTOR.REQUEST_ACTOR_LONG_REST, actor.id)
        }
    }

    func requestInitialActorData() {
        if let actor = SocketManagerWrapper.sharedInstance.actor {
            socket.emit(SocketEvents.IOS.ACTOR.REQUEST_ACTOR_ATTRIBUTES, actor.id)
        }
    }

    func isReady() -> Bool {
        receivedFirstMessage
    }
}
